scripts/generate-ckpt.sh

```diff
@@ -2,30 +2,53 @@
 
 set -e
 
-usage() {
-    echo "Usage: $0 [OPTIONS]"
-    echo ""
-    echo "Options"
-    echo "  --help -h  : Display this message"
-    echo "  -n <n>     : Number of harts"
-    echo "  -b <elf>   : Binary to run in spike"
-    echo "  -p <pc>    : PC to take checkpoint at [default 0x80000000]"
-    echo "  -i <insns> : Instructions after PC to take checkpoint at [default 0]"
-    echo "  -m <isa>   : ISA to pass to spike for checkpoint generation [default rv64gc]"
-    echo "  -o <out>   : Output directory to store the checkpoint in. [default <elf>.<pc>.<insns>.loadarch]"
-    echo "  -r <mem>   : Memory regions to pass to spike. Passed to spike's '-m' flag. [default starting at 0x80000000 with 256MiB]"
-    echo "  -v         : Verbose"
-    exit "$1"
-}
+# assumes that only memory region is at 0x80000000. this might break with more regions (and/or at different locations)
+DEFAULT_MEM_START_ADDR=0x80000000
 
 NHARTS=1
 BINARY=""
-PC="0x80000000"
+PC="$DEFAULT_MEM_START_ADDR"
+INSN=
 INSNS=0
 ISA="rv64gc"
 OUTPATH=""
 MEMOVERRIDE=""
 VERBOSE=0
+DTB=
+DTS=
+TYPE="defaultspikedts"
+
+usage() {
+    echo "Usage: $0 [OPTIONS]"
+    echo ""
+    echo "Misc Options"
+    echo "  --help -h  : Display this message"
+    echo "  -o <out>   : Output directory to store the checkpoint in. [default <elf>.<pc>.<insn>.<insns>.<dtstype>.loadarch]"
+    echo "  -v         : Verbose"
+    echo ""
+    echo "Required Options"
+    echo "  -b <elf>   : Binary to run in spike"
+    echo ""
+    echo "Other Required Options (with defaults if excluded)"
+    echo "  -r <mem>   : Memory regions to pass to spike. Passed to spike's '-m' flag. [default starting at $DEFAULT_MEM_START_ADDR with 256MiB]"
+    echo ""
+    echo "Options Group (one is required)"
+    echo "  -p <pc>    : PC to take checkpoint at [default $PC]"
+    echo "  -t <insn>  : Instruction (in hex) to take checkpoint at [default is to use ${INSN:-none}]"
+    echo "  -i <insns> : Instructions after PC to take checkpoint at [default $INSNS]"
+    echo ""
+    echo "Mutually Exclusive Option Groups (each group is mutually exclusive with one another)"
+    echo "  Group: Use Spike default DTS with modifications (can choose multiple)"
+    echo "    -n <n>     : Number of harts [default $NHARTS]"
+    echo "    -m <isa>   : ISA to pass to spike for checkpoint generation [default $ISA]"
+    echo "  Group: Use custom DT{B,S} (choose one)"
+    echo "    Important Note: a dt{b,s} only affects the devices used, pmps, mmu."
+    echo "                    the isa string and num. of harts from the dt{b,s} is inferred from this script to pass to the spike --isa and -p flags, respectively."
+    echo "    -d <dtb>   : DTB file to use. Passed to spike's '--dtb' flag. [default is to use ${DTB:-none}]"
```

Comment:
We should fix spike to infer the isa correctly from DTS/DTB.

Reply:
What is the reason for supporting both DTS and DTB? I'd like to avoid needless complexity

Reply:
> We should fix spike to infer the isa correctly from DTS/DTB.

Yes, but I'll leave that for a later PR (I need to do other stuff).

> What is the reason for supporting both DTS and DTB? I'd like to avoid needless complexity

This is just for ease of use, we can just support the DTB if you want to reduce flags.

Reply:
If the DTS use case is easier (which I would think would be the case), then we don't have to support the dtb flag.

Reply:
https://github.com/riscv-software-src/riscv-isa-sim/issues/1716